### PR TITLE
SONAR-7464 Optimize NCLOC_DATA and COMMENT_LINES_DATA

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: sonar-scanner-engine/src/main/java/org/sonar/batch/DefaultFileLinesContext.java [hunk 1]
```diff
@@ -115,19 +132,27 @@ public void save() {
...
+          .on(inputFile)
+          .forMetric(metricFinder.findByKey(metricKey))
+          .withValue(data)
+          .save();
         entry.setValue(ImmutableMap.copyOf(lines));
       }
     }
   }
 
+  private Map optimizeStorage(String metricKey, Map<Integer, Object> lines) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/DefaultFileLinesContext.java]: MAJOR Make "optimizeStorage" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-scanner-engine/src/main/java/org/sonar/batch/DefaultFileLinesContextFactory.java [hunk 2]
```diff
@@ -19,34 +19,50 @@
...
+    this.sensorContext = sensorContext;
+    this.metricFinder = metricFinder;
+    this.scannerComponentCache = scannerComponentCache;
+    this.measureCache = measureCache;
   }
 
   @Override
-  public FileLinesContext createFor(Resource model) {
+  public FileLinesContext createFor(Resource resource) {
+    Preconditions.checkArgument(ResourceUtils.isFile(resource));
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/DefaultFileLinesContextFactory.java]: MINOR Remove this use of "ResourceUtils"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-scanner-engine/src/main/java/org/sonar/batch/DefaultFileLinesContextFactory.java [hunk 3]
```diff
@@ -19,34 +19,50 @@
...
     // Reload resource in case it use deprecated key
-    Resource resource = index.getResource(model);
-    return new DefaultFileLinesContext(index, resource);
+    File file = (File) index.getResource(resource);
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/batch/DefaultFileLinesContextFactory.java]: MINOR Remove this use of "File"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
MINOR Remove this use of "File"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 1 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MINOR 3 minor

Watch the comments in this conversation to review them.

**Comment by dbmeneses** [comment]
LGTM

**Pull request merged**